### PR TITLE
Add sane fallback for fee estimation

### PR DESCRIPTION
**Pull request by morcos**
Make fee estimation logic respect GetRequiredFee at all times.
In addition add new commandline option "-fallbackfee" to use when fee estimation does not have sufficient data.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
utACK  cc99c41

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -47,6 +47,12 @@ bool fSendFreeTransactions = DEFAULT_SEND_FREE_TRANSACTIONS;
  * Override with -mintxfee
  */
 CFeeRate CWallet::minTxFee = CFeeRate(DEFAULT_TRANSACTION_MINFEE);
+/**
+ * If fee estimation does not have enough data to provide estimates, the use this fee instead.
```
Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: Nit: `, then use` ...

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -2223,14 +2229,12 @@ CAmount CWallet::GetMinimumFee(unsigned int nTxBytes, unsigned int nConfirmTarge
...
-    }
-    // prevent user from paying a non-sense fee (like 1 satoshi): 0 < fee < minRelayFee
-    if (nFeeNeeded < ::minRelayTxFee.GetFee(nTxBytes))
-        nFeeNeeded = ::minRelayTxFee.GetFee(nTxBytes);
+        // ... unless we don't have enough mempool data for estimatefee, then use fallbackFee
+        if (nFeeNeeded == 0)
+            nFeeNeeded = fallbackFee.GetFee(nTxBytes);
+    }
+    // prevent user from paying a fee below minRelayTxFee or minTxFee
+    nFeeNeeded = std::max(nFeeNeeded, GetRequiredFee(nTxBytes));
```
Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: This is important on it's own. Could you create another atomic commit with just this diff?

File: src/wallet/wallet.h [hunk 3]
```diff
@@ -41,6 +41,8 @@ static const unsigned int DEFAULT_KEYPOOL_SIZE = 100;
 static const CAmount DEFAULT_TRANSACTION_FEE = 0;
 //! -paytxfee will warn if called with a higher fee than this amount (in satoshis) per KB
 static const CAmount nHighTransactionFeeWarning = 0.01 * COIN;
+//! -fallbackfee default
+static const CAmount DEFAULT_FALLBACK_FEE = 20000;
```
Comment (MarcoFalke) [inline src/wallet/wallet.h]: I'd say `11000` is enough, we can determine a "better" `DEFAULT_` for 0.13

**Comment by morcos** [comment]
Addressed @MarcoFalke's comments.
Although left fee = 20k satoshis until the bikeshedding is finished.  I don't care want goes there, but I propose using the result of estimatefee(4) with a significantly longer time decay over the last few months.  I'll post the result of that once calculated.

@wumpus this is meant for 0.12 backport

**Comment by morcos** [comment]
@MarcoFalke ugh.. thanks.  more than just that one fail.  These are probably things that should be fixed b/c the travis tests that fail are implicitly depending on estimatefee returning minrelaytxfee, which isn't a good assumption.  I'm working on it

**Comment by wumpus** [comment]
Reminder: @wumpus on github is not a bitcoin person.

**Comment by morcos** [comment]
Apologies.

@laanwj I think we could use this in 0.12

**Comment by laanwj** [comment]
Concept ACK.

One nit: this adds yet another `-XXXfee` parameter.
Although it makes sense, for the wallet we already have `-mintxfee`, `-paytxfee`, `-maxtxfee` this wild growth of options does get a bit confusing for users, I think.
But I don't know a better solution either.

**Comment by dcousens** [comment]
@laanwj its also not clear whether those parameters are referring to the wallet or to relay policy.

**Comment by btcdrak** [comment]
Concept ACK will test later today.

**Comment by sdaftuar** [comment]
Tested, ACK bebe58b748532157958f9055e4517e280684006c

**Comment by MarcoFalke** [comment]
* utACK 995b9f3
* utACK bebe58b

**Comment by laanwj** [comment]
utACK

**Comment by laanwj** [comment]
Cherry-picked to 0.12 as a36d79bfe247e7fc5c6296fd8603f5094edfe558

**Pull request merged**